File: src/lib/git.ts
import { execaSync } from 'execa';
import { existsSync } from 'fs';
import { join } from 'path';
import { homedir } from 'os';
import type { Barn } from '../types.js';
import { hasValidSshConfig } from './config.js';
import { shellEscape } from './shell.js';

function expandPath(path: string): string {
  if (path.startsWith('~/')) {
    return join(homedir(), path.slice(2));
  }
  return path;
}

export interface GitInfo {
  isGitRepo: boolean;
  remoteUrl?: string;
  branch?: string;
}

export function detectGitInfo(path: string): GitInfo {
  const expandedPath = expandPath(path);

  // Check if .git exists
  if (!existsSync(join(expandedPath, '.git'))) {
    return { isGitRepo: false };
  }

  try {
    // Get remote URL
    const remoteResult = execaSync('git', ['-C', expandedPath, 'config', '--get', 'remote.origin.url'], {
      reject: false,
    });
    const remoteUrl = remoteResult.exitCode === 0 ? remoteResult.stdout.trim() : undefined;

    // Get current branch
    const branchResult = execaSync('git', ['-C', expandedPath, 'rev-parse', '--abbrev-ref', 'HEAD'], {
      reject: false,
    });
    const branch = branchResult.exitCode === 0 ? branchResult.stdout.trim() : undefined;

    return {
      isGitRepo: true,
      remoteUrl,
      branch,
    };
  } catch {
    return { isGitRepo: false };
  }
}

/**
 * Detect git info on a remote server via SSH.
 */
export function detectRemoteGitInfo(path: string, barn: Barn): GitInfo {
  // Verify barn has valid SSH config
  if (!hasValidSshConfig(barn)) {
    return { isGitRepo: false };
  }

  try {
    // Run git commands via SSH
    const result = execaSync('ssh', [
      '-p', String(barn.port),
      '-i', barn.identity_file,
      '-o', 'BatchMode=yes',
      '-o', 'ConnectTimeout=5',
      `${barn.user}@${barn.host}`,
      `cd ${shellEscape(path)} && git config --get remote.origin.url 2>/dev/null && git rev-parse --abbrev-ref HEAD 2>/dev/null`
    ], { timeout: 10000, reject: false });

    if (result.exitCode !== 0) {
      return { isGitRepo: false };
    }

    const lines = result.stdout.trim().split('\n');
    const remoteUrl = lines[0] || undefined;
    const branch = lines[1] || undefined;

    return {
      isGitRepo: true,
      remoteUrl,
      branch,
    };
  } catch {
    return { isGitRepo: false };
  }
}

File: src/git.rs
use std::path::Path;
use std::process::Command;

use crate::types::Barn;

#[derive(Debug, Clone)]
pub struct GitInfo {
    pub is_git_repo: bool,
    pub remote_url: Option<String>,
    pub branch: Option<String>,
}

fn expand_path(path: &str) -> String {
    if let Some(rest) = path.strip_prefix("~/") {
        if let Some(home) = dirs::home_dir() {
            return home.join(rest).to_string_lossy().to_string();
        }
    }
    path.to_string()
}

/// Detect git info for a local path
pub fn detect_git_info(path: &str) -> GitInfo {
    let expanded = expand_path(path);

    // Check if .git exists
    if !Path::new(&expanded).join(".git").exists() {
        return GitInfo { is_git_repo: false, remote_url: None, branch: None };
    }

    let remote_url = Command::new("git")
        .args(["-C", &expanded, "config", "--get", "remote.origin.url"])
        .output()
        .ok()
        .filter(|o| o.status.success())
        .map(|o| String::from_utf8_lossy(&o.stdout).trim().to_string())
        .filter(|s| !s.is_empty());

    let branch = Command::new("git")
        .args(["-C", &expanded, "rev-parse", "--abbrev-ref", "HEAD"])
        .output()
        .ok()
        .filter(|o| o.status.success())
        .map(|o| String::from_utf8_lossy(&o.stdout).trim().to_string())
        .filter(|s| !s.is_empty());

    GitInfo { is_git_repo: true, remote_url, branch }
}

/// Detect git info on a remote server via SSH
pub fn detect_remote_git_info(path: &str, barn: &Barn) -> GitInfo {
    let remote_cmd = format!(
        "cd {} && git config --get remote.origin.url 2>/dev/null && git rev-parse --abbrev-ref HEAD 2>/dev/null",
        crate::tmux::shell_escape(path)
    );

    // BatchMode: detection is a background convenience, so it must fail rather
    // than block on an auth prompt.
    match crate::ssh::run(barn, &remote_cmd, crate::ssh::Opts { batch: true, ..Default::default() }) {
        Ok(stdout) => {
            let lines: Vec<&str> = stdout.trim().lines().collect();
            let remote_url = lines.first().map(|s| s.to_string()).filter(|s| !s.is_empty());
            let branch = lines.get(1).map(|s| s.to_string()).filter(|s| !s.is_empty());
            GitInfo { is_git_repo: true, remote_url, branch }
        }
        Err(_) => GitInfo { is_git_repo: false, remote_url: None, branch: None },
    }
}
